extensions/interactions/NumericExpressionInput/directives/numeric-expression-input-rules.service.ts

```diff
@@ -0,0 +1,117 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Rules service for the NumericExpressionInput interaction.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+import nerdamer from 'nerdamer';
+
+import { MathInteractionsService } from 'services/math-interactions.service.ts';
+import { INumericExpressionAnswer } from 'interactions/answer-defs';
+import { INumericExpressionRuleInputs } from 'interactions/rule-input-defs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class NumericExpressionInputRulesService {
+  MatchesExactlyWith(
+      answer: INumericExpressionAnswer,
+      inputs: INumericExpressionRuleInputs): boolean {
+    let mis = new MathInteractionsService();
+
+    // The expression is first split into terms by addition and subtraction.
+    let answerTerms = mis.getTerms(answer);
+    let inputTerms = mis.getTerms(inputs.x);
+
+    // Now, we try to match all terms between the answer and input.
+    // NOTE: We only need to iterate from the top in the answerTerms list since
+    // in the inputTerms list, we will break the loop each time an element is
+    // removed from it, thus, indexing errors would only arise in the outer
+    // loop.
+    for (let i = answerTerms.length - 1; i >= 0; i--) {
+      for (let j = 0; j < inputTerms.length; j++) {
+        if (mis.termsMatch(answerTerms[i], inputTerms[j])) {
+          answerTerms.splice(i, 1);
+          inputTerms.splice(j, 1);
+          break;
+        }
+      }
+    }
+
+    // The two expressions are considered an exact match if both lists are empty
+    // implying that each term in the answer got uniquely matched with a term
+    // in the input.
+    return answerTerms.length === 0 && inputTerms.length === 0;
+  }
+
+  IsEquivalentTo(
+      answer: INumericExpressionAnswer,
+      inputs: INumericExpressionRuleInputs): boolean {
+    return nerdamer(answer).eq(nerdamer(inputs.x).toString());
+  }
+
+  ContainsSomeOf(
+      answer: INumericExpressionAnswer,
+      inputs: INumericExpressionRuleInputs): boolean {
+    // At least one term should match between answer and input.
+    let mis = new MathInteractionsService();
+
+    // The expression is first split into terms by addition and subtraction.
+    let answerTerms = mis.getTerms(answer);
+    let inputTerms = mis.getTerms(inputs.x);
+
+    for (let answerTerm of answerTerms) {
+      for (let inputTerm of inputTerms) {
+        if (mis.termsMatch(answerTerm, inputTerm)) {
```

Comment:
Can't you just call `termsMatch` with the 2 lists here instead of  on a term by term basis?

Reply:
No, because termsMatch is expecting terms which it will split into parts (by * and /). We still need to split the main expression into terms (by + and -).